feat: implement abstract dataset and refactor Kafka dataset

giddyphysicist opened the pull request:
## Summary
<!-- Describe the changes in this PR. -->
Create new AbstractDataset base class so that new dataset types can inherit from this general class.

Refactor the Kafka dataset to inherit from AbstractDataset

Update node.py and runner.py to handle the more general syntax for the Kafka Dataset.

Update tests

## Motivation
<!-- Describe the motivation for this change. -->
To have a standard way of adding more general dataset types to support types such as batch, streaming, or in-memory datasets.

## Author Checklist

- [x] Changes are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Documentation is updated if necessary.
- [x] Status checks pass.
- [ ] If the version number is changed, it is updated in `pyproject.toml` , `aineko/__init__.py`, `aineko/templates/first_aineko_pipeline/{{cookiecutter.project_slug}}/pyproject.toml`.


## Reviewer Checklist

- [x] Title is accurate and formatted appropriately.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Docs are updated if necessary.
- [x] Code is pythonic and consistent with the rest of the codebase.
- [x] The code is consistent with the [style guide](https://google.github.io/styleguide/pyguide.html).

craycrayfish committed:
feat: initial design

giddyphysicist committed:
chore: utility functions

utility functions for creating producer, consumer, and admin client.

partial implementation of _create, _delete methods

giddyphysicist committed:
chore: docs updates

craycrayfish committed:
feat: design suggestions

giddyphysicist committed:
refactor: move kafka dataset to separate file

giddyphysicist committed:
chore: create status class

giddyphysicist committed:
chore: incremental updates

giddyphysicist committed:
chore: typevar for from_config method

added typevar to fix the from_config method. allows for method to return AbstractDataset or any subclass of AbstractDataset

giddyphysicist committed:
chore: incremental kafka dataset updates

giddyphysicist committed:
chore: node method updates for dataset init

giddyphysicist committed:
fix: inputs

giddyphysicist committed:
chore: lint

giddyphysicist committed:
chore: config for setting up producers/consumers

giddyphysicist committed:
chore: lint

giddyphysicist committed:
feat: create fake inputs and outputs for datasets

giddyphysicist committed:
refactor: use fake datasets for node testing

update from self.consumers --> self.inputs and self.producers --> self.outputs

giddyphysicist committed:
chore: update fixtures for node tests

use new self.inputs and self.outputs syntax

giddyphysicist committed:
test: update FakeDataset tests

giddyphysicist committed:
chore: debug fixes

giddyphysicist committed:
chore: kafka message update

giddyphysicist committed:
chore: websocket_client syntax updates

giddyphysicist committed:
chore: exception handling and docstrings

giddyphysicist committed:
chore: docstrings and pydantic models

giddyphysicist committed:
chore: docstrings

giddyphysicist committed:
chore: lint

giddyphysicist committed:
feat: _initialize method

refactored query layer creation into _initialize method rather than _create method.

Barabazs commented:
Is there a specific reason to move the dataset implementations outside of `aineko.core.dataset`?

What if we went from 
```
├── __init__.py
├── __main__.py
├── cli
│   ├── ...
├── config.py
├── core
│   ├── __init__.py
│   ├── config_loader.py
│   ├── dataset.py (old code?)
│   ├── deploy_config_loader.py
│   ├── node.py
│   ├── node_manager.py
│   └── runner.py
├── datasets
│   ├── __init__.py
│   ├── core.py
│   └── kafka.py
├── ...
...
```

to:
```
├── __init__.py
├── __main__.py
├── cli
│   ├── ...
├── config.py
├── core
│   ├── __init__.py
│   ├── config_loader.py
│   ├── dataset.py (AbstractDataset)
│   ├── deploy_config_loader.py
│   ├── node.py
│   ├── node_manager.py
│   └── runner.py
├── datasets (out of the box Aineko dataset implementations)
│   ├── __init__.py
│   └── kafka.py
├── ...
...
```

Barabazs committed:
refactor: update type hints for python3.8 compatibility

giddyphysicist committed:
chore: exists method and python 3.8 compatibility

giddyphysicist committed:
Merge branch 'feat/abstract-dataset' of https://github.com/aineko-dev/aineko into feat/abstract-dataset

giddyphysicist committed:
refactor: updated fastapi extras node syntax

extended syntax to account for AbstractDatsets as inputs and outputs

giddyphysicist committed:
chore: update fastapi tests

giddyphysicist committed:
chore: lint

giddyphysicist committed:
chore: added 'block' kwarg to _read

giddyphysicist committed:
chore: update unit tests and integration tests

giddyphysicist commented:
> Is there a specific reason to move the dataset implementations outside of `aineko.core.dataset`?
> 
> What if we went from
> 
> ```
> ├── __init__.py
> ├── __main__.py
> ├── cli
> │   ├── ...
> ├── config.py
> ├── core
> │   ├── __init__.py
> │   ├── config_loader.py
> │   ├── dataset.py (old code?)
> │   ├── deploy_config_loader.py
> │   ├── node.py
> │   ├── node_manager.py
> │   └── runner.py
> ├── datasets
> │   ├── __init__.py
> │   ├── core.py
> │   └── kafka.py
> ├── ...
> ...
> ```
> 
> to:
> 
> ```
> ├── __init__.py
> ├── __main__.py
> ├── cli
> │   ├── ...
> ├── config.py
> ├── core
> │   ├── __init__.py
> │   ├── config_loader.py
> │   ├── dataset.py (AbstractDataset)
> │   ├── deploy_config_loader.py
> │   ├── node.py
> │   ├── node_manager.py
> │   └── runner.py
> ├── datasets (out of the box Aineko dataset implementations)
> │   ├── __init__.py
> │   └── kafka.py
> ├── ...
> ...
> ```

Yes I think your suggestion makes more sense -- having the base classes (for both nodes and datasets) under the core directory. Having a separate datasets folder would then make it easier to separate as plugins as well.

giddyphysicist committed:
chore: outline for AbstractAsyncDataset class

giddyphysicist committed:
Revert "chore: outline for AbstractAsyncDataset class"

This reverts commit 722284c2ac8cc189e7ed5fb21fd52b3c4d547afa.

giddyphysicist committed:
chore: core.py moved to core/dataset.py

giddyphysicist committed:
chore: mypy lint style updates

giddyphysicist committed:
chore: test fix inputs

giddyphysicist committed:
chore: update `target` to `location` for dataset config

giddyphysicist committed:
chore: remove print statements

giddyphysicist committed:
chore: update default location and setting

giddyphysicist committed:
chore: update cookiecutter tests

giddyphysicist committed:
chore: remove DatasetConsumer and DatasetProducer from tests

replace with dataset reader

giddyphysicist committed:
chore: added AbstractDataset to aineko __init__

giddyphysicist committed:
chore: generic typing updates

adds more flexibility in function arguments for subclasses

giddyphysicist committed:
chore: docstring updates

giddyphysicist committed:
docs: mkdocs updates

giddyphysicist committed:
chore: vale lint fix

Barabazs commented on aineko/core/dataset.py:
```diff
@@ -1,525 +1,276 @@
+        """Subclass implementation to initialize the dataset query layer."""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _describe(self, *args: T, **kwargs: T) -> str:
+        """Subclass implementation to describe the dataset metadata."""
+        return f"Dataset name: {self.name}"
```
What's the point of this method?

giddyphysicist replied:
This is supposed to be a common place to return metadata about a specific dataset. By default, it just returns the name, but particular implementations can add data to this. It may not be strictly necessary (and if it is going to be a string, we could just make it the `__str__` or `__repr__` method)

Barabazs replied:
Let's remove it and add a `__str__` method instead. (just before the  `from_config` method.
I suggest:
```python
    def __str__(self) -> str:
        """Return the string representation of the dataset."""
        return f"{self.__class__.__name__}({self.name})"
```

<details>
<summary>
Reasoning
 </summary>

* I don't see any benefit in having it and making this part of the abstract dataset.
* It's too broad and unconstrained, which results in a different implementation for each dataset. For example:
  * kafka dataset: https://github.com/aineko-dev/aineko/blob/4544e16ba310661433d57a8d5d08a38a0f78a922/aineko/datasets/kafka.py#L351-L365
  * postgres dataset: https://github.com/aineko-dev/aineko/blob/e59bc2f46a406b1a1c830985206a2b45ecd05b0f/aineko/extras/datasets/postgres.py#L142-L143
* It's always easier to add a method later than to deprecate and remove one.

  </details>

giddyphysicist replied:
this makes more sense. updated this.

Barabazs commented on aineko/core/dataset.py:
```diff
@@ -1,525 +1,276 @@
-    def produce(self, message: Any) -> None:
-        """Stores message in self.values.
+    @classmethod
+    def from_config(cls: Type[A], name: str, config: Dict[str, Any]) -> A:
```
```suggestion
    def from_config(cls, name: str, config: Dict[str, Any]) -> AbstractDataset:
```
cls doesn't need to be type hinted and we know that from_config will return AbstractDataset

giddyphysicist replied:
I get mypy errors when I try implementing it like this (remove Type[A] and change return type to AbstractDataset):
```Undefined variable 'AbstractDataset' (undefined-variable)```

If I try to remove Type[A] and replace the return type as A, I get a different mypy error:
```error: A function returning TypeVar should receive at least one argument containing the same TypeVar  [type-var]```

I think this has to do with from_config returning a subclass of AbstractDataset

giddyphysicist replied:
Oh I see, if I put quotes around AbstractDataset for the return type (and remove Type[A]), it doesn't throw mypy errors. 
I do think what you propose is simpler, but also might have slightly less type safety.

Barabazs commented on aineko/core/dataset.py:
```diff
@@ -1,525 +1,276 @@
+        return f"Dataset name: {self.name}"
+
+    @abc.abstractmethod
+    def _exists(self, *args: T, **kwargs: T) -> bool:
```
Do we actually expect to pass anything to this method?

giddyphysicist replied:
Not for the immediate dataset implementations we have in mind. But I wanted to allow flexibility for other implementations if it is needed. However, if you think it is unnecessary, we could always remove it now and add it back later if it is needed.

Barabazs commented on aineko/core/dataset.py:
```diff
@@ -1,525 +1,276 @@
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _describe(self, *args: T, **kwargs: T) -> str:
```
Do we actually expect to pass anything to this method?

giddyphysicist replied:
Not for our current dataset implementations, but it's possible users who make a different implementation will want something more flexible (like filtering), depending on the structure or how extensive the metadata is.

giddyphysicist committed:
chore: simpler typehint for abstractdataset definition

giddyphysicist committed:
Merge branch 'develop' into feat/abstract-dataset

giddyphysicist committed:
chore: minor fixes

syntax, pydantic models

giddyphysicist committed:
chore: update tests

giddyphysicist committed:
Merge branch 'develop' into feat/abstract-dataset

giddyphysicist commented:
I pushed an update after merging with develop. Note, since the extras were moved to the plugins repo, we will need to update the tests in the plugins repo whenever this PR is merged (compare with tests at commit [6de13ab](https://github.com/aineko-dev/aineko/pull/150/commits/6de13ab7ae4a222edeef9f854859f81aea2f88f0) for reference for how to update the plugins tests)

Barabazs commented on aineko/datasets/kafka.py:
```diff
@@ -0,0 +1,839 @@
+
+
+# pylint: enable=too-few-public-methods
+# pylint: disable=unused-argument
```
```suggestion
# pylint: enable=too-few-public-methods
# pylint: disable=unused-argument
```
The first disable rule is obsolete afaik and the unused-argument rule should be handled in the method itself.  (See paragraph about unused arguments here: https://google.github.io/styleguide/pyguide.html#214-decision)

So in the `read` method you would add something like
`del timeout  # unused because this is a fake dataset` to the first line after the docstring.

Barabazs commented on aineko/datasets/kafka.py:
I notice a lot of inconsistent use of Consumer/Producer vs Input/Output. (both here as in other files)

giddyphysicist replied:
I made a pass through to update this, thanks

Barabazs commented on aineko/models/config_schema.py:
```diff
@@ -16,6 +16,7 @@ class Dataset(BaseModel):
             """Dataset model."""
 
             type: str
+            location: Optional[str]
```
```suggestion
            location: Optional[str] = None
```

Barabazs commented on aineko/datasets/kafka.py:
```diff
@@ -0,0 +1,839 @@
+        Args:
+            message: message to produce.
+        """
+        self.values.append(message)
```
This isn't consistent with the new dataset implementation. This should either be merged into one "FakeKafka" or (preferably) be handled in the main Kafka class.

giddyphysicist replied:
updated with a combined class FakeKafka

Barabazs commented on mkdocs.yml:
```diff
@@ -102,12 +102,12 @@ nav:
+          - 'api_reference/kafka_dataset.md'
+          - 'api_reference/config.md'
+          - 'api_reference/config_loader.md'
+          - 'api_reference/runner.md'
```
The quotation marks are not needed here. This adds extra diffs to the git history and is also inconsistent with the rest of the file.

Barabazs commented on tests/conftest.py:
```diff
@@ -74,8 +74,10 @@ class DummyNode(AbstractNode):
-            msg = self.consumers["input"].consume(how="next", timeout=0)
-            self.producers["output"].produce(msg)
+            print(self.inputs)
+            print(self.outputs)
```
```suggestion
```
I suppose this was used while developing, but this doesn't add value to the tests.

Barabazs commented on tests/integration/test_runner_with_kafka.py:
```diff
@@ -123,21 +136,35 @@ def test_write_read_to_kafka(start_service, subtests):
+            count_messages = dataset.consume_all(end_message="END")
+            print("count_messages are...")
+            print(count_messages)
+            print("...")
```
```suggestion
```

Barabazs commented on tests/integration/test_runner_with_kafka.py:
```diff
@@ -123,21 +136,35 @@ def test_write_read_to_kafka(start_service, subtests):
+            print("count_messages are...")
+            print(count_messages)
+            print("...")
+            # assert count_messages[0]["source_pipeline"] == "integration_test_read"
```
Why is this commented out?

giddyphysicist replied:
i added it back in. originally, i wasn't going to implement since we're updating the message structure to remove metadata. but i think that's something that should wait until the dataset validation PR

Barabazs replied:
Yes, we shouldn't modify anything that's out of scope for this feature.

Barabazs commented on aineko/datasets/__init__.py:
Have you considered importing the Kafka dataset here, so we can use it as aineko.datasets.Kafka ?

giddyphysicist replied:
I hadn't considered it. Do you think this would have a large benefit?

Barabazs replied:
Imo it could, but maybe it's worth waiting until we have more dataset types and reassess. I was mainly thinking about how it could look like for the pipeline configs:
```diff
- type: aineko.datasets.kafka.KafkaDataset
+ type: aineko.datasets.KafkaDataset
```

But the aineko-plugins won't be able to support that since the first 2 names are namespace packages... (example `aineko_plugins.datasets.postgres.PostgresDataset`)

Barabazs commented on aineko/__init__.py:
```diff
@@ -9,12 +9,7 @@
-    FakeDatasetConsumer,
-    FakeDatasetProducer,
-)
+from aineko.core.dataset import AbstractDataset
```
I'm not sure why we would want to expose this at the root level of the package. A regular user has no use for this and an advanced user that wants to implement a custom dataset will know where to look by reading the docs.

giddyphysicist replied:
I was mainly thinking about the symmetry between AbstractNode and AbstractDataset. Also, we access the AbstractDataset.from_config method from the pipeline Runner as a generalized way of loading in subclasses of AbstractDataset. I'm ok with removing it from here, but it seems to balance with AbstractNode

Barabazs replied:
> I was mainly thinking about the symmetry between AbstractNode and AbstractDataset.

I'm not sure I understand what you mean by this.

The AbstractNode is used by the user when creating their own node. And every user needs to create a node, so it makes sense to expose it at the root level. But that's not the case for the AbstractDataset.

> Also, we access the AbstractDataset.from_config method from the pipeline Runner

True, but that doesn't require us to expose it at this level.

giddyphysicist replied:
That sounds fine. I updated.

Barabazs commented on aineko/datasets/kafka.py:
```diff
@@ -0,0 +1,839 @@
+    dataset_config: Optional[Dict[str, Any]] = {}
+
+
+class Kafka(AbstractDataset):
```
I think we should rename this to `KafkaDataset`. Kafka is too ambiguous and generic, whereas KafkaDataset is very clear and specific to what the class is about.

giddyphysicist replied:
I had originally called it KafkaDataset, but I think @craycrayfish preferred to call it something simpler. I'm fine either way, but would lean towards KafkaDataset

Barabazs replied:
Let's go with KafkaDataset.

giddyphysicist replied:
Sounds good. I updated it to KafkaDataset

Barabazs commented on tests/core/test_dataset.py:
The file structure of the tests should match the actual file structure as much as possible. This should live under `/tests/datasets/test_kafka.py`.

giddyphysicist replied:
updated

Barabazs commented on tests/core/test_dataset.py:
The new KafkaDataset implementation does not have enough tests. Passing the existing tests is not enough when we add/modify that many features.

giddyphysicist replied:
Thanks. I've added a few tests, and I think the refactored original tests cover a good bit of the functionality. Are there any particular places you recommend more testing?

Barabazs replied:
Thanks! It seems like all the methods and features are tested (some indirectly). 🚀

Barabazs commented on tests/conf/integration_test_empty_dataset.yml:
```diff
@@ -16,6 +16,8 @@ pipeline:
     messages:
-      type: kafka_stream
+      type: aineko.datasets.kafka.Kafka
+      location: localhost:9092
```
We should have a few tests where location is missing, location is missing and not in ENV vars, etc.

giddyphysicist replied:
Good call, I added a few tests in.

Barabazs commented on aineko/core/dataset.py:
```diff
@@ -1,529 +1,278 @@
+            raise
+        except Exception as e:
+            message = f"Failed to read dataset {self.name}."
+            raise DatasetError(message) from e
```
All methods which raise an exception should mention it in the docstring.

giddyphysicist replied:
Added.

Barabazs commented on aineko/core/dataset.py:
What's the rationale for implementing the public methods as wrappers for the abstract internal methods? Why not define the required public methods as abstract methods instead?

giddyphysicist replied:
I think this was mainly to add generic error handling (and potentially more future features like logging or telemetry) as a standard wrapper. This also similarly follows the format for `execute` wrapping user-implemented  `_execute` method for AbstractNode subclasses

Barabazs replied:
I don't think this pattern is useful or worth the trade offs. The most important downside (imo) is that we're hiding the useful information from the user. They don't know what a method does, which arguments to pass and what it's returning:

<img width="541" alt="image" src="https://github.com/aineko-dev/aineko/assets/31799121/5d78ed50-9c30-4508-b841-c826c6a2ee05">

vs

<img width="851" alt="image" src="https://github.com/aineko-dev/aineko/assets/31799121/025d1bb6-3302-4fed-a127-76d736f7764f">

I would suggest removing the wrapper methods and making the public methods abstract methods.

giddyphysicist replied:
Yes, I think this makes more sense. Especially with the method details. I updated.

Barabazs commented on aineko/core/node.py:
```diff
@@ -118,36 +126,66 @@ def setup_datasets(
             has_pipeline_prefix: whether the dataset name has pipeline name
                 prefix
+
+        TODO: update docstring
```
TODO?

giddyphysicist replied:
forgot to remove this, thanks

Barabazs commented on aineko/core/node.py:
```diff
@@ -118,36 +126,66 @@ def setup_datasets(
         )
+        # topics have already been created from runner.
+
+        # now create the producers and consumers connections:
```
```suggestion
```

Barabazs commented on aineko/core/node.py:
```diff
@@ -208,9 +248,10 @@ def log(self, message: str, level: str = "info") -> None:
-        )
+        # self.producers[DEFAULT_KAFKA_CONFIG.get("LOGGING_DATASET")].produce(
+        #     out_msg
+        # )
```
To be deleted?

giddyphysicist replied:
deleted

Barabazs commented on aineko/core/node.py:
```diff
@@ -358,15 +399,20 @@ def run_test_yield(
+            # if self.consumers and all(
+            #     consumer.empty for consumer in self.consumers.values()
+            # ):
+            #     run_loop = False
```
?

giddyphysicist replied:
old code, i removed

Barabazs commented on aineko/core/runner.py:
```diff
@@ -139,88 +139,65 @@ def prepare_datasets(
-        # Create all configured datasets
-        datasets = kafka_client.create_topics(dataset_list)
+            # update dataset config here:
+            dataset: AbstractDataset = AbstractDataset.from_config(
+                dataset_name, dataset_config
+            )
+            datasets.append(dataset)
+
+        # Create logging dataset
+        logging_config = {
+            "location": "localhost:9092",
+            "type": "aineko.datasets.kafka.Kafka",
+        }
+        logging_dataset_name = DEFAULT_KAFKA_CONFIG.get("LOGGING_DATASET")
+        logger.info(
+            "Creating dataset: %s: %s", logging_dataset_name, logging_config
+        )
+        logging_dataset: AbstractDataset = AbstractDataset.from_config(
+            logging_dataset_name, logging_config
+        )
```
The type hints don't seem correct. The datasets are not AbstractDataset in this context.

giddyphysicist replied:
This one is a bit tricky. They are all technically subclasses of AbstractDataset, which can by implied by the syntax. They aren't all necessarily a single type of subclass (like KafkaDataset) since we want the runner to load the most general case of datasets. I couldn't find a recommended way of handling this, other than referencing the generic superclass of AbstractDataset

Barabazs replied:
Thanks, I see how that makes sense.

Barabazs commented on aineko/core/node.py:
```diff
@@ -65,8 +66,15 @@ class AbstractNode(ABC):
+        name (str): name of the node
+        pipeline_name (str): name of the pipeline
+        params (dict): dict of parameters to be used by the node
         consumers (dict): dict of DatasetConsumer objects for inputs to node
         producers (dict): dict of DatasetProducer objects for outputs of node
```
Are these still attributes?

giddyphysicist replied:
no, i removed it.

Barabazs commented:
Please mark the PR as "ready for review" when I should revisit. Thanks

giddyphysicist committed:
chore: clearer pylint for FakeDatasetInput

giddyphysicist committed:
style: remove quotes from mkdocs yaml

giddyphysicist committed:
chore: default value for location

giddyphysicist committed:
chore: remove print statements from tests

giddyphysicist committed:
chore: source_node and source_pipeline attributes

added in attributes for message metadata (will likely be removed later when message structure is streamlined, but provided here for backwards format compatibility)

giddyphysicist committed:
chore: remove old comments

giddyphysicist committed:
chore: docstring fix remove consumers producers

giddyphysicist committed:
chore: docstring cleanup

giddyphysicist committed:
chore: combined test readers and writers

combined into fakekafka

giddyphysicist committed:
chore: tests for default `location` value

giddyphysicist committed:
chore: lint and docstrings

giddyphysicist committed:
chore: change name to KafkaDataset

giddyphysicist committed:
chore: update terminology

consume--> read
produce --> write
consumers --> inputs
producers --> outputs

craycrayfish commented:
> Is there a specific reason to move the dataset implementations outside of `aineko.core.dataset`?
> 
> What if we went from
> 
> ```
> ├── __init__.py
> ├── __main__.py
> ├── cli
> │   ├── ...
> ├── config.py
> ├── core
> │   ├── __init__.py
> │   ├── config_loader.py
> │   ├── dataset.py (old code?)
> │   ├── deploy_config_loader.py
> │   ├── node.py
> │   ├── node_manager.py
> │   └── runner.py
> ├── datasets
> │   ├── __init__.py
> │   ├── core.py
> │   └── kafka.py
> ├── ...
> ...
> ```
> 
> to:
> 
> ```
> ├── __init__.py
> ├── __main__.py
> ├── cli
> │   ├── ...
> ├── config.py
> ├── core
> │   ├── __init__.py
> │   ├── config_loader.py
> │   ├── dataset.py (AbstractDataset)
> │   ├── deploy_config_loader.py
> │   ├── node.py
> │   ├── node_manager.py
> │   └── runner.py
> ├── datasets (out of the box Aineko dataset implementations)
> │   ├── __init__.py
> │   └── kafka.py
> ├── ...
> ...
> ```

The main reason I proposed this was because I imagine Aineko to have several decoupled components in the future - runner, datasets, config loader, deployment config. And they wold have their own directories eventually. So it makes sense to start decoupling them

Barabazs commented on docs/developer_guide/node_implementation.md:
```diff
@@ -63,20 +63,20 @@ class MySumNode(AbstractNode):
+### Inputs & Outputs
 
-Node classes inherit attributes named `self.producers` and `self.consumers` that are each a dictionary, with keys being the dataset name and values being `DatasetProducer` and `DatasetConsumer` objects respectively. These objects allow you to produce/consume data to/from a dataset.
+Node classes inherit attributes named `self.inputs` and `self.outputs` that are each a dictionary, with keys being the dataset name and values being subclasses of `AbstractDataset`. These objects allow you to read/write data to/from a dataset.
```
```suggestion
Node classes inherit attributes named `self.inputs` and `self.outputs` that are each a dictionary, with keys being the dataset name and values being subclasses of `AbstractDataset`. These objects allow you to read/write data from/to a dataset.
```
You switched the order, so it should also be updated int rest of the sentence.

giddyphysicist replied:
thanks, updated

Barabazs commented on docs/developer_guide/node_implementation.md:
```diff
@@ -91,78 +91,78 @@ class MySumNode(AbstractNode):
 
-!!! warning "Producers and Consumers must be included in the pipeline configuration"
+!!! warning "Inputs and Outputs must be included in the pipeline configuration"
     They must be defined in the `outputs` and `inputs` list respectively to be available to the node. If a dataset is not available in a Node's catalog, a `KeyError` will be raised.
```
```suggestion
!!! warning "Inputs and Outputs must be included in the pipeline configuration"
    They must be defined in the `inputs` and `outputs` list respectively to be available to the node. If a dataset is not available in a Node's catalog, a `KeyError` will be raised.
```

giddyphysicist replied:
updated

Barabazs commented on docs/index.md:
```diff
@@ -99,8 +99,8 @@ Supercharge your development by using pre-built nodes directly in your pipeline.
-- [`DatasetConsumer`](./api_reference/dataset_consumer.md)
-- [`DatasetProducer`](./api_reference/dataset_producer.md)
+- [`AbstractDataset`](./api_reference/abstract_dataset.md)
+- [`Kafka`](./api_reference/kafka_dataset.md)
```
```suggestion
- [`KafkaDataset`](./api_reference/kafka_dataset.md)
```

giddyphysicist replied:
thanks, updated this

Barabazs commented on tests/integration/test_runner_with_kafka.py:
```diff
@@ -102,19 +102,30 @@ def test_write_read_to_kafka(start_service, subtests):
         try:
             runner.run()
         except ray.exceptions.RayActorError:
```
```suggestion
        except ray.exceptions.RayActorError:
            # This is expected because we activated the poison pill
            pass
        ....
```
You should keep the same structure when doing this try-except block with a PoisonPill activation. (see comments in a previous PR: https://github.com/aineko-dev/aineko/pull/112#pullrequestreview-1794510059)

giddyphysicist replied:
Yes, that's more consistent. I updated.

Barabazs commented on docs/index.md:
Commenting here since I can't comment on an unmodified file:
The `docs/examples/aineko_dream.md` still uses the old notation of `self.consumers` etc

giddyphysicist replied:
updated.

Barabazs reviewed (changes requested):
Thank you for addressing the feedback. There are still some issues to resolve, but we're getting closer to merging.

Barabazs reviewed (commented):
I think the kafka related parts are still too tightly intertwined with the rest of the codebase. (examples are node or runner) We should definitely revisit this in a new PR.

Barabazs commented:
@giddyphysicist can you respond to the feedback conversations once it's addressed instead of resolving it please? Otherwise it becomes challenging to follow up properly if everything is marked as resolved.

giddyphysicist committed:
chore: add __str__ method and remove _describe

giddyphysicist committed:
chore: remove AbstractDataset from root level of package

giddyphysicist committed:
chore: update aineko dream docs to new syntax

giddyphysicist committed:
chore: word order and docs name update

giddyphysicist committed:
chore: try except poison pill order update

giddyphysicist committed:
chore: remove wrapper functions for abstract dataset

Barabazs reviewed (approved):
This is a huge step forwards! Thank you  🚀 🔥

Barabazs commented on aineko/core/dataset.py:
````diff
@@ -1,529 +1,213 @@
+        def _delete(self, **kwargs) -> Any:
+            pass
 
+        def _initialize(self, **kwargs) -> Any:
+            pass
 
-class FakeDatasetProducer:
-    """Fake dataset producer for testing purposes.
+        def _exists(self, **kwargs) -> bool:
+            pass
+    ```
 
-    The class will store in its state the list of values produced
-    by the node.
+    If `MyDataset` was defined in the file
+    `./aineko/datasets/mydataset.py`, a new dataset
+    can be created using the `from_config` method:
 
-    Args:
-        dataset_name: name of the dataset
-        node_name: name of the node that is producing the dataset
+    Example:
+    ```python
 
-    Attributes:
-        dataset_name (str): name of the dataset
-        node_name (str): name of the node that is producing the dataset
-        values (list): list of mock data values produced by the node
+    dataset = AbstractDataset.from_config(
+        name="my_dataset_instance",
+        config={
+            "type": "aineko.datasets.mydataset.MyDataset",
+            "location": "foo",
+            "params": {
+                "param_1": "bar"
+            }
+        }
+    )
+    ```
     """
````
I just noticed that the docstring still mentions the method names with an underscore. This should be updated before merging please.

Barabazs replied:
Updated it in  dc6552b98fc52a4cff6c3fcb05e7b8cab24c7210

Barabazs committed:
doc: update docstrings to remove the underscore

Barabazs committed:
doc: fix import statement for AbstractDataset

craycrayfish commented on aineko/core/node.py:
```diff
@@ -120,35 +122,61 @@ def setup_datasets(
             }
         )
 
+        for dataset_name in inputs:
+            if self.inputs[dataset_name].type == "kafka":
+                consumer_params = ConsumerParams(
+                    **{
+                        "dataset_name": dataset_name,
+                        "node_name": self.name,
+                        "pipeline_name": self.pipeline_name,
+                        "prefix": prefix,
+                        "has_pipeline_prefix": has_pipeline_prefix,
+                        "consumer_config": DEFAULT_KAFKA_CONFIG.get(
+                            "CONSUMER_CONFIG"
+                        ),
+                    }
+                )
+                self.inputs[dataset_name].initialize(
+                    create_consumer=True, connection_params=consumer_params
+                )
```
This needs to be a separate function so we can initialize different dataset types here (and in the runner). This is fine if it will be refactored in a subsequent PR

craycrayfish commented on aineko/datasets/kafka.py:
````diff
@@ -0,0 +1,823 @@
+
+
+class ConsumerParams(BaseModel):
+    """Parameters for initializing a Kafka Consumer.
+
+    Passed in as connection_params when calling
+        ```python
+        self.initialize(create_consumer=True,
+                      connection_params=ConsumerParams(...))
+        ```
+    """
+
+    dataset_name: str
+    node_name: str
+    pipeline_name: str
+    prefix: Optional[str] = None
+    has_pipeline_prefix: bool
+    consumer_config: Dict[str, Any] = DEFAULT_KAFKA_CONFIG.get(
+        "CONSUMER_CONFIG"
+    )
+
+
+class ProducerParams(BaseModel):
+    """Parameters for initializing a Kafka Producer.
+
+    Passed in as conection_params when calling
+        ```python
+        self.initialize(create_producer=True,
+                      connection_params=ProducerParams(...))
+        ```
+    """
+
+    dataset_name: str
+    node_name: str
+    pipeline_name: str
+    prefix: Optional[str] = None
+    has_pipeline_prefix: bool
+    producer_config: Dict[str, Any] = DEFAULT_KAFKA_CONFIG.get(
+        "PRODUCER_CONFIG"
+    )
````
Can create a generic KafkaParams class which both of these inherit so we don't need to repeat the params

giddyphysicist replied:
That seems cleaner, thanks.

craycrayfish commented on aineko/datasets/kafka.py:
```diff
@@ -0,0 +1,823 @@
+    def initialize(
+        self,
+        connection_params: Union[ConsumerParams, ProducerParams],
+        create_consumer: bool = False,
+        create_producer: bool = False,
```
why not set the arg as create: Literal["consumer", "producer"] so that the fact that you can choose one is implicit

Barabazs replied:
Maybe reader/writer to be consistent with the terms were using now?

giddyphysicist replied:
Yes, this seems better. I updated.

craycrayfish commented:
Amazing work! Thank you for the push!n Added a couple of non-blocking comments

giddyphysicist committed:
chore: base class KafkaParams and create literal

giddyphysicist committed:
Merge branch 'develop' into feat/abstract-dataset
